ios-sockets/SwiftWebSockets/SwiftWebSockets/Networking/NWWebSocket.swift

```diff
@@ -0,0 +1,226 @@
+//
+//  NWWebSocket.swift
+//  SwiftWebSockets
+//
+//  Created by Dan Browne on 21/09/2020.
+//  Copyright © 2020 Neas Lease. All rights reserved.
+//
+
+import Foundation
+import Network
+
+open class NWWebSocket: WebSocketConnection {
+
+    // MARK: - Public properties
+
+    weak var delegate: WebSocketConnectionDelegate?
+
+    // MARK: - Private properties
+
+    private let connection: NWConnection
+    private let endpoint: NWEndpoint
+    private let parameters: NWParameters
+    private var pingTimer: Timer?
+
+    // MARK: - Initialization
+
+    init(request: URLRequest, connectAutomatically: Bool = false) {
+
+        endpoint = .url(request.url!)
+
+        if request.url?.port == 80 {
+            parameters = NWParameters.tcp
+        } else {
+            parameters = NWParameters.tls
+        }
+
+        let wsOptions = NWProtocolWebSocket.Options()
+        wsOptions.autoReplyPing = true
+        parameters.defaultProtocolStack.applicationProtocols.insert(wsOptions, at: 0)
+
+        connection = NWConnection(to: endpoint, using: parameters)
+
+        if connectAutomatically {
+            connect()
+        }
+    }
+
+    init(url: URL, connectAutomatically: Bool = false) {
+
+        endpoint = .url(url)
+
+        if url.port == 80 {
+            parameters = NWParameters.tcp
+        } else {
+            parameters = NWParameters.tls
+        }
+
+        let wsOptions = NWProtocolWebSocket.Options()
+        wsOptions.autoReplyPing = true
+        parameters.defaultProtocolStack.applicationProtocols.insert(wsOptions, at: 0)
+
+        connection = NWConnection(to: endpoint, using: parameters)
+
+        if connectAutomatically {
+            connect()
+        }
+    }
+
+    // MARK: - WebSocketConnection conformance
+
+    func connect() {
+        connection.stateUpdateHandler = stateDidChange(to:)
+        listen()
+        connection.start(queue: .main)
```

Comment:
Random question, do you think this has to happen on the main queue?

Reply:
I've just added a `connectionQueue` parameter with a default value to the initialisers after testing this will work off the main queue as well :)